server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java

```diff
@@ -2677,6 +2677,9 @@ private String orchestrateExtractVolume(long volumeId, long zoneId) {
         }
         // perform extraction
         ImageStoreEntity secStore = (ImageStoreEntity)dataStoreMgr.getImageStore(zoneId);
+        if (secStore == null) {
+            throw new InvalidParameterValueException(String.format("Secondary storage to satisfy storage needs cannot be found for zone: %d", zoneId));
+        }
         String value = _configDao.getValue(Config.CopyVolumeWait.toString());
```

Comment:
What about the `VolumeVO uploadVolume` and the `GetUploadParamsResponse uploadVolume` , that also use  `getImageStore` , will you also handle the null returns there?

Reply:
Not seeing any reference for getImageStore(Long zoneId) here,
https://github.com/shapeblue/cloudstack/blob/secondary-storage-limit-3287/server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java#L290-L388
`DataStore getImageStore(String storeUuid, Long zoneId)` from `TemplateManager.java` used in these methods already handles null value.

Reply:
Yes, you are correct, hadn't seen the `CloudRuntimeException` inside the method.